File: software/examples/javascript/ExampleModbusSlave.js
var Tinkerforge = require('tinkerforge');

var HOST = 'localhost';
var PORT = 4223;
var UID = 'XYZ'; // Change XYZ to the UID of your RS485 Bricklet

var ipcon = new Tinkerforge.IPConnection(); // Create IP connection
var rs485 = new Tinkerforge.BrickletRS485(UID, ipcon); // Create device object

ipcon.connect(HOST, PORT,
    function (error) {
        console.log('Error: ' + error);
    }
); // Connect to brickd
// Don't use device before ipcon is connected

ipcon.on(Tinkerforge.IPConnection.CALLBACK_CONNECTED,
    function (connectReason) {
        // Set operating mode to Modbus RTU slave
        rs485.setMode(Tinkerforge.BrickletRS485.MODE_MODBUS_SLAVE_RTU);

        // Modbus specific configuration:
        // - slave address = 17
        // - master request timeout = 0ms (unused in slave mode)
        rs485.setModbusConfiguration(17, 0);
    }
);

// Register Modbus slave write single register request callback
rs485.on(Tinkerforge.BrickletRS485.CALLBACK_MODBUS_SLAVE_WRITE_SINGLE_REGISTER_REQUEST,
    // Callback function for Modbus slave write single register request callback
    function (requestID, registerAddress, registerValue) {
        console.log('Request ID: ' + requestID);
        console.log('Register Address: ' + registerAddress);
        console.log('Register Value: ' + registerValue);

        if (registerAddress != 42) {
            console.log('Requested invalid register address');
            rs485.modbusSlaveReportException(requestID, Tinkerforge.BrickletRS485.EXCEPTION_CODE_ILLEGAL_DATA_ADDRESS);
        } else {
            rs485.modbusSlaveAnswerWriteSingleRegisterRequest(requestID);
        }
    }
);

console.log('Press key to exit');
process.stdin.on('data',
    function (data) {
        ipcon.disconnect();
        process.exit(0);
    }
);

File: software/examples/perl/example_modbus_slave.pl
#!/usr/bin/perl

use Tinkerforge::IPConnection;
use Tinkerforge::BrickletRS485;

use constant HOST => 'localhost';
use constant PORT => 4223;
use constant UID => 'XYZ'; # Change XYZ to the UID of your RS485 Bricklet

my $ipcon = Tinkerforge::IPConnection->new(); # Create IP connection
our $rs485 = Tinkerforge::BrickletRS485->new(&UID, $ipcon); # Create device object

# Callback subroutine for Modbus slave write single register request callback
sub cb_modbus_slave_write_single_register_request
{
    my ($request_id, $register_address, $register_value) = @_;

    print "Request ID: $request_id\n";
    print "Register Address: $register_address\n";
    print "Register Value: $register_value\n";

    if($register_address != 42)
    {
        print "Error: Invalid register address\n";
        $rs485->modbus_slave_report_exception($request_id, $rs485->EXCEPTION_CODE_ILLEGAL_DATA_ADDRESS);
    }
    else
    {
        $rs485->modbus_slave_answer_write_single_register_request($request_id);
    }
}

$ipcon->connect(&HOST, &PORT); # Connect to brickd
# Don't use device before ipcon is connected

# Set operating mode to Modbus RTU slave
$rs485->set_mode($rs485->MODE_MODBUS_SLAVE_RTU);

# Modbus specific configuration:
# - slave address = 17
# - master request timeout = 0ms (unused in slave mode)
$rs485->set_modbus_configuration(17, 0);

# Register Modbus slave write single register request callback to subroutine
# cb_modbus_slave_write_single_register_request
$rs485->register_callback($rs485->CALLBACK_MODBUS_SLAVE_WRITE_SINGLE_REGISTER_REQUEST,
                          'cb_modbus_slave_write_single_register_request');

print "Press key to exit\n";
<STDIN>;
$ipcon->disconnect();
